casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala

```diff
@@ -96,50 +91,32 @@ object DagOperations {
   }
 
   /**
-    * Conceptually, the GCA is the first point at which the histories of b1 and b2 diverge.
-    * We compute by finding the first block from genesis for which there exists a child of that block which is an ancestor of b1 or b2 but not both.
-    *
-    * TODO: Implement a GCA that doesn't require the genesis block (see https://rchain.atlassian.net/browse/RCHAIN-3002)
-    * TODO: Remove usage of BlockStore by just using BlockDagRepresentation (see https://rchain.atlassian.net/browse/RCHAIN-3003)
+    * Conceptually, the LCA is the lowest point at which the histories of b1 and b2 diverge.
+    * We compute by finding the first block that is the "oldest" (has highest blocknum) common block.
     */
-  def greatestCommonAncestorF[F[_]: Monad: BlockStore](
-      b1: BlockMessage,
-      b2: BlockMessage,
-      genesis: BlockMessage,
+  def lowestCommonAncestorF[F[_]: Monad](
+      b1: BlockMetadata,
+      b2: BlockMetadata,
       dag: BlockDagRepresentation[F]
-  ): F[BlockMessage] =
+  ): F[BlockMetadata] = {
+    type BlockWithDepth = (Long, BlockMetadata)
+    implicit val blockNumDecreasing: Ordering[BlockWithDepth] =
+      (l: BlockWithDepth, r: BlockWithDepth) => r._1.compare(l._1)
+
+    def getParents(p: BlockMetadata): F[Set[BlockMetadata]] =
+      p.parents.map(v => dag.lookup(v)).sequence.map(_.toSet.flatten)
+
+    def extractParentsFromHighestNumBlock(s: SortedSet[BlockWithDepth]): F[SortedSet[BlockWithDepth]] = {
+      val (head, tail) = (s.head._2, s.tail)
+      getParents(head).map(nb => tail ++ nb.map(v => (v.blockNum, v)))
```

Comment:
Is difficult to understand what `nb => tail ++ ...` does. Please extract it as a inner named function for improved readability.